Doc/tutorial/errors.rst

```diff
@@ -496,3 +496,153 @@ used in a way that ensures they are always cleaned up promptly and correctly. ::
+     |   File "<stdin>", line 3, in f
+     | ExceptionGroup: there were problems
+     +-+---------------- 1 ----------------
+       | OSError: error 1
+       +---------------- 2 ----------------
+       | SystemError: error 2
+       +------------------------------------
+   >>> try:
+   ...     f()
+   ... except Exception as e:
+   ...     print(f'caught {type(e)}: e')
+   ...
+   caught <class 'ExceptionGroup'>: e
+   >>>
+
+By using ``except*`` instead of ``except``, we can selectively
+handle only the exceptions in the group that match a certain
+type. In the following example, which shows a nested exception
+group, each ``except*`` clause extracts from the group exceptions
+of a certain type while letting all other exceptions propagate to
+other clauses and eventually to be reraised. ::
+
+   >>> def f():
+   ...     raise ExceptionGroup("group1",
+   ...                          [OSError(1),
+   ...                           SystemError(2),
+   ...                           ExceptionGroup("group2",
+   ...                                          [OSError(3), RecursionError(4)])])
+   ...
+   >>> try:
+   ...     f()
+   ... except* OSError as e:
+   ...     print("There were OSErrors")
+   ... except* SystemError as e:
+   ...     print("There were SystemErrors")
+   ...
+   There were OSErrors
+   There were SystemErrors
+     + Exception Group Traceback (most recent call last):
+     |   File "<stdin>", line 2, in <module>
+     |   File "<stdin>", line 2, in f
+     | ExceptionGroup: group1
+     +-+---------------- 1 ----------------
+       | ExceptionGroup: group2
+       +-+---------------- 1 ----------------
+         | RecursionError: 4
+         +------------------------------------
+   >>>
+
+Note that the exceptions nested in an exception group must be instances,
+not types. This is because in practice the exceptions would typically
+be ones that have already been raised and caught by the program, along
+the following pattern::
+
+   >>> excs = []
+   ... for test in tests:
+   ...     try:
+   ...         test.run()
+   ...     except Exception as e:
+   ...         excs.append(e)
+   ...
+   ... if excs:
+   ...    raise ExceptionGroup("Test Failures", excs)
+   ...
+
+
+Enriching Exceptions with Notes
+===============================
+
+When an exception is created in order to be raised, it is usually initialized
+with information that describes the error that has occurred. There are cases
+where it is useful to add information after the exception was caught. For this
+purpose, exceptions have a mutable field ``__note__`` that can be assigned to
+a string which is included in formatted tracebacks. ::
+
+   >>> try:
+   ...     raise TypeError('bad type')
+   ... except Exception as e:
+   ...     e.__note__ = 'Add some information'
+   ...     raise
+   ...
+   Traceback (most recent call last):
+     File "<stdin>", line 2, in <module>
+   TypeError: bad type
+   Add some information
```

Comment:
It feels like traceback printing should be more explicit about separating the note from the exception type.  Maybe something like:

```
Traceback (most recent call last):
  File "/tmp/foo.py", line 2, in <module>
    raise TypeError('bad type')
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: bad type
Note: More information
```

OTOH, I guess you can always do that in your own note.  But then I think that having a single `__note__` which can only take a string will mean that if the exception is caught and notarized at different levels of the code, you're going to run into contention over the value of this attribute.  This is essentially the case now with `args` although that's messier because `args` is underspecified for historical reasons.  More discussion about the semantics and use cases of `__note__` seems worthwhile.

Reply:
I thought of adding a "Note:" prefix but then decided to make it more free form. Multi-line notes are also supported, see https://github.com/python/cpython/blob/e9898bf153d26059261ffef11f7643ae991e2a4c/Lib/test/test_traceback.py#L1657

So if you want vertical whitespace you can just start with a couple of newlines.

Reply:
Zac's PEP should address this.

Reply:
@Zac-HD

Reply:
It's actually already free form:

```
>>> class E(Exception):
    def __str__(self):
        return '''
             OOOOOOOOOOO
         OOOOOOOOOOOOOOOOOOO
      OOOOOO  OOOOOOOOO  OOOOOO
    OOOOOO      OOOOO      OOOOOO
  OOOOOOOO  #   OOOOO  #   OOOOOOOO
 OOOOOOOOOO    OOOOOOO    OOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOO  OOOOOOOOOOOOOOOOOOOOOOOOO  OOOO
 OOOO  OOOOOOOOOOOOOOOOOOOOOOO  OOOO
  OOOO   OOOOOOOOOOOOOOOOOOOO  OOOO
    OOOOO   OOOOOOOOOOOOOOO   OOOO
      OOOOOO   OOOOOOOOO   OOOOOO
         OOOOOO         OOOOOO
             OOOOOOOOOOOO

'''
... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... 
>>> 
>>> raise E()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
E: 
             OOOOOOOOOOO
         OOOOOOOOOOOOOOOOOOO
      OOOOOO  OOOOOOOOO  OOOOOO
    OOOOOO      OOOOO      OOOOOO
  OOOOOOOO  #   OOOOO  #   OOOOOOOO
 OOOOOOOOOO    OOOOOOO    OOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOO  OOOOOOOOOOOOOOOOOOOOOOOOO  OOOO
 OOOO  OOOOOOOOOOOOOOOOOOOOOOO  OOOO
  OOOO   OOOOOOOOOOOOOOOOOOOO  OOOO
    OOOOO   OOOOOOOOOOOOOOO   OOOO
      OOOOOO   OOOOOOOOO   OOOOOO
         OOOOOO         OOOOOO
             OOOOOOOOOOOO

```

Reply:
I'm keen to stick with freeform, not least because Hypothesis might have *several* multi-line notes to attach (as a single string, of course).  The obvious case is where we want to provide the minimal failing example, *and* point out certain lines at which control-flow diverged for failing examples.  For example:

```python
def fake_complicated_calculation(n):
    if n % 16 == 1:
        # If we run this line, the test will fail with a confusing
        # error - and this function won't be in the traceback!
        # Hypothesis' explain phase points it out though :-)
        return "oops, bad data"
    return n // 5

from hypothesis import Phase, given, settings, strategies as st

@given(st.integers(min_value=0))
@settings(phases=[Phase.generate, Phase.shrink, Phase.explain], derandomize=True)
def test(x):
    result = fake_complicated_calculation(x)
    assert (result % 2) in {0, 1}
    assert x < 1e6, "just to force an ExceptionGroup"

test()
```
```
$ python3.11 example.py

  + Exception Group Traceback (most recent call last):
  |       ...
  | ExceptionGroup: Hypothesis found 2 distinct failures.
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "example.py", line 15, in test
    |     assert (result % 2) in {0, 1}
    |             ~~~~~~~^~~
    | TypeError: not all arguments converted during string formatting
    |
    | Falsifying example: test(
    |     x=1,
    | )
    | Explanation:
    |     These lines were always and only run by failing examples:
    |         example.py:6
    +---------------- 2 ----------------
    | Traceback (most recent call last):
    |   ...
    | AssertionError: just to force an ExceptionGroup
    | ...
    +------------------------------------
```

Reply:
> Hypothesis might have several multi-line notes to attach (as a single string, of course)

Do you worry about complicated string hacking in the case of multiple notes being added at different places?

Reply:
I think this is likely to be rare and easy to handle if you care to; the PEP (currently!) says:

> Assigning a new string value overrides an existing note; if concatenation is desired users are responsible for implementing it with e.g.::
>
>     e.__note__ = msg if e.__note__ is None else e.__note__ + "\n" + msg

Reply:
Complicated or not, is string hacking the right thing to do? And how much can Hypothesis assume about `__note__`?
Consider this silly example, which could be refactored in various ways:
```
    def test_silly(self):
        for a in range(10):
            with self.subTest(a=a):
                for b in range(10):
                    with self.subTest(b=b):
                        self.assertNotEqual(a*b, 12)
 ```
Today's unittest has its own way of handling multiple exceptions, but if `subTest` it was being written today,  would `__note__` not be a good fit here?
If it would, how would `subTest` be expected to work, and how should it interoperate with Hypothesis?

Reply:
I've replied over on https://github.com/python/peps/pull/2201#issuecomment-999355383